tests/testGncOptimizer.cpp

```diff
@@ -0,0 +1,625 @@
+/* ----------------------------------------------------------------------------
+
+ * GTSAM Copyright 2010, Georgia Tech Research Corporation,
+ * Atlanta, Georgia 30332-0415
+ * All Rights Reserved
+ * Authors: Frank Dellaert, et al. (see THANKS for the full author list)
+
+ * See LICENSE for the license information
+
+ * -------------------------------------------------------------------------- */
+
+/**
+ * @file    testGncOptimizer.cpp
+ * @brief   Unit tests for GncOptimizer class
+ * @author  Jingnan Shi
+ * @author  Luca Carlone
+ * @author  Frank Dellaert
+ *
+ * Implementation of the paper: Yang, Antonante, Tzoumas, Carlone, "Graduated Non-Convexity for Robust Spatial Perception:
+ * From Non-Minimal Solvers to Global Outlier Rejection", RAL, 2020. (arxiv version: https://arxiv.org/pdf/1909.08605.pdf)
+ */
+
+#include <gtsam/slam/dataset.h>
+
+#include <gtsam/nonlinear/LevenbergMarquardtOptimizer.h>
+#include <gtsam/nonlinear/GaussNewtonOptimizer.h>
+#include <gtsam/nonlinear/NonlinearFactorGraph.h>
+#include <tests/smallExample.h>
+
+#include <CppUnitLite/TestHarness.h>
+
+using namespace std;
+using namespace gtsam;
+
+using symbol_shorthand::X;
+using symbol_shorthand::L;
+static double tol = 1e-7;
+
+/* ************************************************************************* */
+template<class BaseOptimizerParameters>
+class GncParams {
+public:
+  /** Verbosity levels */
+  enum VerbosityGNC {
+    SILENT = 0, SUMMARY, VALUES
+  };
+
+  /** Choice of robust loss function for GNC */
+  enum RobustLossType {
+    GM /*Geman McClure*/, TLS /*Truncated least squares*/
+  };
+
+  using BaseOptimizer = GaussNewtonOptimizer; // BaseOptimizerParameters::OptimizerType;
```

Comment:
Oh ?

Reply:
yes, this exactly the problem we need to solve. I can do:
using BaseOptimizer = BaseOptimizerParameters::OptimizerType;
Then I can go to LevenbergMarquardtParams and define OptimizerType as LevenbergMarquardtOptimizer.
However, LevenbergMarquardtParams does not know about LevenbergMarquardtOptimizer, so I'm not sure how to make that define work. did you have something different in mind?

Reply:
I think you need to modify LevenbergMarquardtParams to add the typedef there

Reply:
ok I'll try that, but I suspect an easier way would be to template on the Optimizer instead of the Params.